### ISSUE TITLE
multi-valued vs. time series measurements

### ISSUE THREAD

**Issue by nedmsmith** [ietf-rats-wg/draft-ietf-rats-corim#24]
reference, endorsed, and reference-endorsed triples support the idea that the object of the triple can be multi-valued. There are two use cases for its use. (a) when there are multiple of the same type measurements. For example, 'digest' or 'raw-value-group' where evidence is constructed so that the first digest is in the first array position and the second digest is in the second array position. Otherwise, the digests are not named (that is they share the same environment-map identity).

In the time series use case there is a single evidence value that is compared across multiple reference values. Reference values are ordered according to the series semantics (e.g., time) such that the measurement at position 0 is compared first, then the measurement at position 1 and so forth. 

The verifier is undecided which comparison algorithm to apply unless the triple contains extra information. It is possible to use mkey to supply the extra information, but mkey (as part of measurement-map) should expect to match an mkey in Evidence as it becomes part of the evidence matching logic. Additionally, if mkey is deprecated this approach is no longer available to disambiguate the use cases. 

A solution should include, at least, a boolean that indicates whether to apply time series vs. multi-valued attributes semantics. 
It is best to keep it independent from the environment-map and the measurement-map structures as these carry matching semantics overhead.

Possible solutions:
(a) Add a flag for series:
```
reference-triple-record = [
  environment-map
  [ + measurement-map ]
]
```

```
measurement-map = {
  ? &(series: 0) => bool .default false
  &(mval: 1) => measurement-values-map
  ? &(authorized-by: 2) => [ + $crypto-key-type-choice ]
}
```

(b) Add another level of array:
```
measurement-map = {
  &(mval: 1) => [ + measurement-values-map ] ; If multiple array entries exist then apply series comparison semantics
  ? &(authorized-by: 2) => [ + $crypto-key-type-choice ]
}
```
(c) Don't deprecate mkey:
```
measurement-map = {
  ? &(mkey: 0) => $measured-element-type-choice
  &(mval: 1) => measurement-values-map
  ? &(authorized-by: 2) => [ + $crypto-key-type-choice ]
}
$measured-element-type-choice /= bool .default true ; only use the flag when series semantics are desired
```

**Comment by shnwc** [comment]
There is another possible solution, to have an optional dedicated second list of measurement-maps dedicated for time series:

(d) Dedicated time series list:

reference-triple-record = [
  environment-map
  [ * measurement-map ] ; multi-valued list, which can be an empty list
  ? [ + measurement-map ] ; If this second list is present, it contains a time series of measurement-maps
]

If each entry in a time series should be a multi-valued list, then the definition will be:

reference-triple-record = [
  environment-map
  [ * measurement-map ] ; multi-valued list, which can be an empty list
  ? [ + [ + measurement-map] ] ; If this second list is present, it contains a time series of measurement-map lists.
]

**Comment by nedmsmith** [comment]
Issue #24 should be considered in context with issue #2 as both conditional endorsements and time series use cases can exist simultaneously.

**Comment by nedmsmith** [comment]
After today's discussion, people are leaning toward option (b) - adding a nested array that contains time series and relying on array indies for sequencing. However, the exact triple definition should incorporate thinking from issue #2.

**Comment by shnwc** [comment]
Option (b) sounds fine. With issue #2 incorporated, will the triple definitions be something like below?

```
$$triples-map-extension //= (
  comid.reference-endorsed-triples => [ + reference-endorsed-triple-record ]
)
comid.reference-endorsed-triples = 7

reference-endorsed-triple-record = [
  environment-map ; identifies the target environment
  [ + reference-endorsed-measurement-map ] ; reference values with endorsed values
]

reference-endorsed-measurement-map = {
  &(mval: 1) => [ + measurement-values-map ] ; If multiple array entries exist then apply series comparison semantics
  ? &(authorized-by: 2) => [ + $crypto-key-type-choice ]
}

reference-endorsed-measurement-map = {
  &(ref-vals :0) => [ +  measurement-values-map ]
  &(end-vals :1) => [ + measurement-values-map ]
}
```

**Comment by nedmsmith** [comment]
There’s a definition similar to that in a branch but it lacks ref/end-Val’s as array for time series use case.

**Comment by nedmsmith** [comment]
Another refinement round.

The triple _subject_ is a compound structure consisting of an environment identifier plus measurements that, when matched to evidence, is a condition of the _object_.
The _object_ is a time series of conditional reference and endorsed triples
```
conditional-reference-series-triple-record = [
  [ 
    environment-map,
    measurement-map / measurement-values-map
  ] ; triple subject is a compound statement
  [ + reference-endorsed-record ] ; triple object is a series
]
 
reference-endorsed-record = [
  measurement-map / measurement-values-map  ; reference values
  ? measurement-map / measurement-values-map  ; conditionally endorsed values, could be empty, but still a valid expression in a series
]
```

**Comment by nedmsmith** [comment]
Resolved by PR #36 - closing issue.
